internal/tofu/node_resource_abstract_instance.go

```diff
@@ -105,6 +108,84 @@ func (n *NodeAbstractResourceInstance) References() []*addrs.Reference {
 	return nil
 }
 
+func (n *NodeAbstractResourceInstance) ResolveProvider(ctx EvalContext) tfdiags.Diagnostics {
+	var diags tfdiags.Diagnostics
+
+	log.Printf("[TRACE] Resolving provider key for %s", n.Addr)
+
+	if n.Config != nil && n.Config.ProviderConfigRef != nil && n.Config.ProviderConfigRef.KeyExpression != nil {
+		if n.ResolvedProviderKeyExpr != nil {
+			// Module key and resource key are required. This is not allowed!
+			diags = diags.Append(fmt.Errorf("TODO better message: provider key required in both module and resource"))
+			return diags
+		}
+
+		keyExpr := n.Config.ProviderConfigRef.KeyExpression
+		keyData := ctx.InstanceExpander().GetResourceInstanceRepetitionData(n.Addr)
+		keyScope := ctx.EvaluationScope(nil, nil, keyData)
+
+		keyVal, keyDiags := keyScope.EvalExpr(keyExpr, cty.DynamicPseudoType)
+		diags = diags.Append(keyDiags)
+		if keyDiags.HasErrors() {
+			return diags
+		}
+
+		if keyVal.Type() == cty.String {
+			n.ResolvedProviderKey = addrs.StringKey(keyVal.AsString())
+			log.Printf("[TRACE] Resource %s used provider key %q", n.Addr, keyVal.AsString())
+		} else if keyVal.Type() == cty.Number {
+			var intVal int
+			gocty.FromCtyValue(keyVal, &intVal)
+			n.ResolvedProviderKey = addrs.IntKey(intVal)
+			log.Printf("[TRACE] Resource %s used provider key %v", n.Addr, intVal)
+		} else {
+			panic("Bad key type")
+		}
+	} else if n.ResolvedProviderKeyExpr != nil {
+		// This is a hack, but not as bad as what is comment out below
+		moduleInstanceForKey := n.Addr.Module[:len(n.ResolvedProviderKeyPath)]
+
+		keyExpr := n.ResolvedProviderKeyExpr
+		keyData := ctx.InstanceExpander().GetModuleInstanceRepetitionData(moduleInstanceForKey)
+		keyScope := ctx.WithPath(moduleInstanceForKey).EvaluationScope(nil, nil, keyData)
+
+		keyVal, keyDiags := keyScope.EvalExpr(keyExpr, cty.DynamicPseudoType)
+		diags = diags.Append(keyDiags)
+		if keyDiags.HasErrors() {
+			return diags
+		}
+
+		if keyVal.Type() == cty.String {
```

Comment:
It's rare in the OpenTofu language to require an exact type like this, as opposed to trying to convert to an expected target type.

Typically I'd recommend first deciding which type is expected based on the declarations and then trying to convert to that type, but it seems like that would be tricky in this case because we don't have access to the configuration in here and so we can't, for example, check whether the provider configuration block we're referring to uses `count`, `for_each`, or neither. :thinking: 

Unfortunately just trying to convert to each type in turn is probably not a good idea because all numbers can convert to strings and some strings can convert to numbers.

Maybe requiring an exact type is okay for now, but that _will_ make this behave subtly differently to how the index operator works in other situations. If we can find some way to get enough config-related information here to _know_ what kind of key the provider configuration is expecting then that'd be the most ideal way to solve this, but I dunno how feasible that is with how things are currently shaped here.

Reply:
[846c114](https://github.com/opentofu/opentofu/pull/2105/commits/846c114ae67fb59ca0ed2dd1629a716da6c7dae5)

Reply:
Interesting to remember that `addrs.ParseInstanceKey` was already requiring exact types!

The other existing uses of `addrs.ParseInstanceKey` are are exclusively for the `cty.Value` coming from a `hcl.TraverseIndex` though, and that's guaranteed to always be either `cty.String` or `cty.Number` because the expression itself is required to be a constant in those cases.

I think this does still remain the only place where a _dynamically-generated_ instance key is required to be of exactly the right type rather than only to convertible to the appropriate type. For example, HCL's own indexing logic uses the type of the collection to decide what key type is needed and then converts to that type: https://github.com/hashicorp/hcl/blob/d20d07fa73707e58954493bdd600c1fb0a0d4c77/ops.go#L58-L82

Since we've removed support for `count` from this initial implementation anyway, there currently cannot possibly be any provider configuration that would need numeric instance keys. Therefore we could choose to just unconditionally convert to `cty.String` for now, with a note to do something better later if we decide to introduce `count` support. (The "something better" would hopefully be to actually check what kind of instance key the provider configuration is _supposed_ to have based on declaration, and then convert to that the same way that HCL indexing does.)

Reply:
I think this can be fixed when we refactor/built out the full provider address reference system.